File: backend/src/main/java/com/team2/auctionality/model/Product.java
package com.team2.auctionality.model;

import com.team2.auctionality.enums.ProductStatus;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "product")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(name = "title")
    private String title;

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private ProductStatus status;

    @Column(name = "start_price")
    private Float startPrice;

    @Column(name = "current_price")
    private Float currentPrice;

    @Column(name = "buy_now_price")
    private Float buyNowPrice;

    @Column(name = "bid_increment")
    private Float bidIncrement;

    @Column(name = "start_time")
    private LocalDateTime startTime;

    @Column(name = "end_time")
    private LocalDateTime endTime;

    @Column(name = "auto_extension_enabled")
    private Boolean autoExtensionEnabled;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
    }

    // ==== RELATIONS ====

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private User seller;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @org.hibernate.annotations.BatchSize(size = 20)
    private List<Bid> bids;

    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @org.hibernate.annotations.BatchSize(size = 20)
    private List<ProductImage> images;
}

File: frontend/src/interfaces/Product.ts
import type Category from "./Category";

export interface ProductImage {
  id: number;
  url: string;
  isThumbnail: boolean;
  productId: number;
}

export interface SellerInfo {
  id: number;
  fullName: string | null;
  avatarUrl: string | null;
  ratingPercent: number | null;
}

export interface HighestBidderInfo {
  id: number;
  maskedName: string | null;
  avatarUrl: string | null;
  ratingPercent: number | null;
}

export interface Product {
  id: number;
  title: string;
  status: string;
  startPrice?: number | null;
  currentPrice?: number | null;
  buyNowPrice?: number | null;
  bidIncrement?: number | null;
  startTime?: string | null; // ISO string from backend
  endTime?: string | null; // ISO string from backend
  autoExtensionEnabled: boolean;
  description?: string | null;
  createdAt?: string | null; // ISO string from backend
  sellerId: number;
  sellerInfo?: SellerInfo | null;
  highestBidderInfo?: HighestBidderInfo | null;
  category?: Category | null;
  images?: ProductImage[];
  bidCount?: number; // For top products
}

export interface ProductImageRequest {
  url: string;
  isThumbnail: boolean;
}

export interface ProductRequest {
  id: number | null;
  title: string;
  status: string;
  startPrice: number | null;
  buyNowPrice?: number | null;
  bidIncrement: number | null;
  startTime?: Date | null; // ISO string
  endTime?: Date | null; // ISO string
  autoExtensionEnabled: boolean;
  sellerId: number;
  categoryId: number | null;
  subcategoryId?: number | null;
  description?: string;
  images?: ProductImageRequest[];
}
